sql: refactor plan node control flow

DrewKimball opened the pull request:
#### sql: change SetZoneConfig to return type Ack

This commit changes the return type for `tree.SetZoneConfig` to `Ack`
from `RowsAffected`. This mirrors similar statements, and the row count
has been broken (always zero) ever since this statement was implemented
in the declarative schema changer, anyway. Removing the need to track
row count allows a few related functions to be simplified.

Epic: None

Release note: None

#### sql: remove unnecessary nodes and interfaces related to mutation planNodes

This commit adds two new helper structs for use in mutation nodes and
other planNodes that need to output the number of affected rows:
`mutationOutputHelper` and `rowsAffectedOutputHelper`. The former is
now embedded into the mutation planNodes apart from `deleteRangeNode`,
which embeds `rowsAffectedOutputHelper` instead.

This commit also removes `spoolNode`, `serializeNode`, and `rowCount`
node, with `mutationOutputHelper` replacing the work they used to do.
This also allowed the removal of the `batchedPlanNode` interface and
its methods.

These changes simplify some factory logic, and should simplify the
process of moving from `planNode` implementations to processors in
the future.

This change also fixes #139378, where the `spoolNode` added for a mutation
with a RETURNING clause would break some of the EXPLAIN ANALYZE stats.

Fixes #139378

Release note (bug fix): Previously, CockroachDB would omit execution
statistics in EXPLAIN ANALYZE output for mutation nodes when a RETURNING
clause was used. The bug has been present since before v21.1 and is now fixed.

#### sql: remove planNodeFastPath interface for ops that return rows affected

Previously, there were two ways that `planNode` implementations would
propagate the number of rows affected (if using that output mode):
1. Implementing the `planNodeFastPath`, which provided a method to
   directly pull the cached row count.
2. A special loop in `planNodeToRowSource`, which simply called `Next()`
   repeatedly and counted the number of iterations.

This commit takes advantage of work done in the previous commit for
operators that need to return the number of rows affected. Now, the
helper structs from the previous commit are used to track this count
internally, and then return the final count through the usual `Next()`
and `Values()` calls. This allows for the removal of special logic
in `planNodeToRowSource`, as well as the `planNodeFastPath` interface.

As with the previous commit, this one seeks to remove departures from
the usual `Start()`, `Next()`, and `Values()` control flow, to make
future refactors easier.

Epic: None

Release note: None

DrewKimball committed:
sql: remove planNodeFastPath interface for ops that return rows affected

Previously, there were two ways that `planNode` implementations would
propagate the number of rows affected (if using that output mode):
1. Implementing the `planNodeFastPath`, which provided a method to
   directly pull the cached row count.
2. A special loop in `planNodeToRowSource`, which simply called `Next()`
   repeatedly and counted the number of iterations.

This commit takes advantage of work done in the previous commit for
operators that need to return the number of rows affected. Now, the
helper structs from the previous commit are used to track this count
internally, and then return the final count through the usual `Next()`
and `Values()` calls. This allows for the removal of special logic
in `planNodeToRowSource`, as well as the `planNodeFastPath` interface.

As with the previous commit, this one seeks to remove departures from
the usual `Start()`, `Next()`, and `Values()` control flow, to make
future refactors easier.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/145934)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
Nice cleanup! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 24 of 24 files at r1, 22 of 22 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-:-OPWng44-v0l1A7JIVFQ:b-jy7zf3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/delete_swap.go` line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWiJ061kaydK9j2TS2:-OPWiJ061kaydK9j2TS3:by3isuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/delete_swap.go#L105)):*
> ```Go
> 		// all secondary indexes.
> 		if condErr := (*kvpb.ConditionFailedError)(nil); errors.As(err, &condErr) {
> 			// Reset the table writer so that it looks like there were no rows to
> ```

nit: this comment is now stale.
___
*[`pkg/sql/plan_node_output_helper_test.go` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWiiC7CYZrhpRpd1di:-OPWiiC7CYZrhpRpd1dj:b-tlugx8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/plan_node_output_helper_test.go#L24)):*
> ```Go
> 
> func TestMutationOutputHelper(t *testing.T) {
> 	rnd, _ := randutil.NewPseudoRand()
> ```

nit: I think we usually use `NewTestRand` in tests.
___
*[`pkg/sql/update_swap.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWiC416zQqmtAYsCLd:-OPWiC416zQqmtAYsCLe:b-wya00e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/update_swap.go#L51)):*
> ```Go
> 	}
> 
> 	// Run the mutation to completion. DeleteSwap only processes one row, so no
> ```

nit: `s/DeleteSwap/UpdateSwap/`.
___
*[`pkg/sql/insert.go` line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWh0pM-cHdMyOB6EbQ:-OPWh0pM-cHdMyOB6EbR:b-sw9svb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/insert.go#L250)):*
> ```Go
> 	}
> 
> 	// If result rows need to be accumulated, do it.
> ```

nit: `s/If result rows .../Result rows .../`.
___
*[`pkg/sql/distsql_physical_planner.go` line 499 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWgbQ--1gkyGJSWiSM:-OPWgbQ--1gkyGJSWiSN:b-i8eiph) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/distsql_physical_planner.go#L499)):*
> ```Go
> }
> 
> func shouldWrapPlanNodeForExecStats(planCtx *PlanningCtx, node planNode) bool {
> ```

nit: we should now inline this method in its only call site.
___
*[`pkg/sql/cancel_queries.go` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWm0f6B9sR4o6TZc8J:-OPWm0f6B9sR4o6TZc8K:bp005cq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f444ed5b4af82f9013dde21369ba6567091cc01/pkg/sql/cancel_queries.go#L20)):*
> ```Go
> )
> 
> type cancelQueriesNode struct {
> ```

nit: why did you decide to also modify `cancelQueriesNode` and `cancelSessionsNode`? These didn't implement `planNodeFastPath` interface. Is it because of the new assertion added in this commit?
___
*[`pkg/sql/opt_exec_factory.go` line 1219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWhc4w7LL8PkzGUC3P:-OPWhc4w7LL8PkzGUC3Q:b-9a1ptm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b5994cb063e494fa7013c731170e5f539d9ba96/pkg/sql/opt_exec_factory.go#L1219)):*
> ```Go
> 	flags exec.PlanFlags,
> ) (exec.Plan, error) {
> 	// No need to spool at the root.
> ```

This removal should fix #139378. If that's the case, we should include a release note with some regression tests (I have a couple in https://github.com/cockroachdb/cockroach/pull/145207/commits/5e395fbbe60f81b86e0f8a89f8d6e27adc25ac8a).


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 5 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145934)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/insert.go` line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPqmyFe52rWBKUXhZxn:-OPqmyFe52rWBKUXhZxo:bilkmi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/insert.go#L274)):*
> ```Go
> 	}
> 
> 	// Run the mutation to completion.
> ```

Does this pattern change the behavior? Previously rows would be pulled through the query plan via `BatchedNext` on-demand, right? Now we're running through all runs immediately and accumulating them all in a rows container. Prior to this change, would we return some rows to the client and clear the container, using less memory for mutations returning many rows? Even rows weren't flushed to the client, they might be flushed to a buffer in a CTE, no?
___
*[`pkg/sql/delete.go` line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPqlwLUD-X4dXEKXXnV:-OPqlwLUD-X4dXEKXXnW:bbcrdjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/delete.go#L112)):*
> ```Go
> 	for {
> 		if err = params.p.cancelChecker.Check(); err != nil {
> 			return lastBatch, err
> ```

nit: Is there a need to return `true` in error cases? If not, using a `false` constant is more clear and idiomatic IMO.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-:-OVQHCew70wEMGZbwc7U:bz5mair)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/cancel_queries.go` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWm0f6B9sR4o6TZc8J:-OVQDlLb-N4CBGpft_Sc:bf43hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f444ed5b4af82f9013dde21369ba6567091cc01/pkg/sql/cancel_queries.go#L20)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why did you decide to also modify `cancelQueriesNode` and `cancelSessionsNode`? These didn't implement `planNodeFastPath` interface. Is it because of the new assertion added in this commit?
</blockquote></details>

There was still a "fast path" for row-count type nodes even if they didn't implement `planNodeFastPath`. The return type for those nodes is [RowsAffected](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/sem/tree/stmt.go#L760-L761), which would previously cause it via [this check](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/distsql_physical_planner.go#L4279) to return a row count [here](https://github.com/cockroachdb/cockroach/blob/7a03efe422d8d728158a0280c043adf46102d3b3/pkg/sql/plan_node_to_row_source.go#L194).

By making this change, we can safely get rid of the special logic in `planNodeToRowSource`.
___
*[`pkg/sql/delete.go` line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPqlwLUD-X4dXEKXXnV:-OVPiUs3AER7eL85b7Va:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/delete.go#L112)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Is there a need to return `true` in error cases? If not, using a `false` constant is more clear and idiomatic IMO.
</blockquote></details>

Done.
___
*[`pkg/sql/delete_swap.go` line 105 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWiJ061kaydK9j2TS2:-OVPihvf9tSzr78X3KOl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/delete_swap.go#L105)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this comment is now stale.
</blockquote></details>

Done.
___
*[`pkg/sql/distsql_physical_planner.go` line 499 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWgbQ--1gkyGJSWiSM:-OVPj2Lz6_Emra8BkAOq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/distsql_physical_planner.go#L499)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we should now inline this method in its only call site.
</blockquote></details>

Done.
___
*[`pkg/sql/insert.go` line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWh0pM-cHdMyOB6EbQ:-OVPjHNL1nNdHMP0Fz6k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/insert.go#L250)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/If result rows .../Result rows .../`.
</blockquote></details>

Done.
___
*[`pkg/sql/insert.go` line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPqmyFe52rWBKUXhZxn:-OVPcOJtAL9jczEsYg7s:bi8l2fk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/insert.go#L274)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Does this pattern change the behavior? Previously rows would be pulled through the query plan via `BatchedNext` on-demand, right? Now we're running through all runs immediately and accumulating them all in a rows container. Prior to this change, would we return some rows to the client and clear the container, using less memory for mutations returning many rows? Even rows weren't flushed to the client, they might be flushed to a buffer in a CTE, no?
</blockquote></details>

In the [old code path](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/opt_exec_factory.go#L1923-L1934), there were two cases:
1. We need rows from the mutation node, in which case we add a [spoolNode](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/spool.go#L16-L21) on top of it.
2. We don't need rows, in which case we add a [rowCountNode](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/plan_batch.go#L160-L172) on top of the mutation.

Both types of "wrapper" nodes ensured that the mutation is executed to completion upon the call to `startExec`, which happens before any rows are returned to the caller.
___
*[`pkg/sql/opt_exec_factory.go` line 1219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWhc4w7LL8PkzGUC3P:-OVQGmw-94Q3GHNGWkDP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b5994cb063e494fa7013c731170e5f539d9ba96/pkg/sql/opt_exec_factory.go#L1219)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This removal should fix #139378. If that's the case, we should include a release note with some regression tests (I have a couple in https://github.com/cockroachdb/cockroach/pull/145207/commits/5e395fbbe60f81b86e0f8a89f8d6e27adc25ac8a).
</blockquote></details>

Done.
___
*[`pkg/sql/update_swap.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWiC416zQqmtAYsCLd:-OVPds-4CV-UrTZlN4eN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/update_swap.go#L51)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/DeleteSwap/UpdateSwap/`.
</blockquote></details>

Done.
___
*[`pkg/sql/plan_node_output_helper_test.go` line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWiiC7CYZrhpRpd1di:-OVPdySyCX8VNItBJTsE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/plan_node_output_helper_test.go#L24)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think we usually use `NewTestRand` in tests.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 29 of 29 files at r3, 23 of 23 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145934)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mgartner)

___
*[`pkg/sql/cancel_queries.go` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWm0f6B9sR4o6TZc8J:-OWNvDWCDkzzzWSvbojJ:b-kufkx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f444ed5b4af82f9013dde21369ba6567091cc01/pkg/sql/cancel_queries.go#L20)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

There was still a "fast path" for row-count type nodes even if they didn't implement `planNodeFastPath`. The return type for those nodes is [RowsAffected](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/sem/tree/stmt.go#L760-L761), which would previously cause it via [this check](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/distsql_physical_planner.go#L4279) to return a row count [here](https://github.com/cockroachdb/cockroach/blob/7a03efe422d8d728158a0280c043adf46102d3b3/pkg/sql/plan_node_to_row_source.go#L194).

By making this change, we can safely get rid of the special logic in `planNodeToRowSource`.
</blockquote></details>

I see, thanks.
___
*[`-- commits` line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNvXsPBm9IjIB-2gbB:-OWNvXsPBm9IjIB-2gbC:b-waqdgd):*
nit: technically, the first commit fixes the issue, no?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/explain_analyze` line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNn1IKACuBXnuGNWLi:-OWNn1IKACuBXnuGNWLj:b-19658l) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bce8048de0d4911c864d83ac29dd621a7c0ecce/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze#L207)):*
> ```
> 
> # Regression test for not showing execution stats for mutation nodes when they
> # are wrapped in a spoolNode (which is the case when RETURNING is present)
> ```

nit: this comment is now stale, maybe reword it as "Regression test for now showing execution stats for mutation nodes with RETURNING clause."?
___
*[`pkg/sql/colexec/columnarizer.go` line 220 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNxNRp3tGpqpyYfZj6:-OWNxNRp3tGpqpyYfZj7:b-i14hqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bce8048de0d4911c864d83ac29dd621a7c0ecce/pkg/sql/colexec/columnarizer.go#L220)):*
> ```Go
> 
> // IsColumnarizerAroundRowsAffectedNode returns true if the provided Operator is a
> // Columnarizer that has "fast-path" enabled planNodeToRowSource wrapper as the
> ```

nit: something like `s/"fast-path" enabled planNodeToRowSource wrapper/planNodeToRowSource wrapper in "rows affected" mode/`.
___
*[`pkg/sql/plan_node_to_row_source.go` line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNwhZLCV_jh3YKzXnV:-OWNwhZLCV_jh3YKzXnW:brra9bx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bce8048de0d4911c864d83ac29dd621a7c0ecce/pkg/sql/plan_node_to_row_source.go#L74)):*
> ```Go
> 	if p.rowsAffected {
> 		// The node returns a single integer value with the number of rows affected.
> 		p.outputTypes = []*types.T{types.Int}
> ```

nit: I think it should be safe to use a global singleton for the single INT type to avoid an allocation here. Perhaps leave a TODO to look into this.
___
*[`pkg/sql/tablewriter.go` line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNtzDsAFYMAcNEu43M:-OWNtzDsAFYMAcNEu43N:bjhghix) ([raw file](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/tablewriter.go#L224)):*
> ```Go
> 
> func (tb *tableWriterBase) clearLastBatch(ctx context.Context) {
> 	tb.lastBatchSize = 0
> ```

nit: I think we can now remove `lastBatchSize` field?
___
*[`pkg/sql/delete.go` line 198 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNryPdDKkr_jzb-YCo:-OWNryPdDKkr_jzb-YCp:b-6e50lo) ([raw file](https://github.com/cockroachdb/cockroach/blob/70d2bfe5d60f388e79fe1be6f86c9c73a7e70277/pkg/sql/delete.go#L198)):*
> ```Go
> 		return err
> 	}
> 	// NOTE: we intentionally do not increment the modified row count if there
> ```

I'm a bit confused about this setup. `r.td.row` will include the necessary KV requests to perform the deletion into `r.td.b`, but that KV batch will not be sent yet to the KV layer, which happens only in `deleteSwapNode.BatchedNext` via `td.finalize` call, so we shouldn't be seeing the ConditionFailedError here if it occurs (the only errors we should be seeing here are the encoding ones).

I think if the final swap PR is picked up on top of this change, the tests there would fail.

Perhaps just remove this comment and reset the counter to zero where we previously modified `rowsWritten`?
___
*[`pkg/sql/delete_swap.go` line 106 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNt2P970SNR483dV6d:-OWNt2P970SNR483dV6e:b-igrv3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/delete_swap.go#L106)):*
> ```Go
> 			// Reset the table writer so that it looks like there were no rows to
> 			// delete.
> 			d.run.td.rowsWritten = 0
> ```

nit: just double-checking: `rowsWritten` will now only be used for ensuring that we don't exceed `rowsWrittenLimit`? Perhaps, it'd be good to highlight that fact somehow? Maybe a comment, or extracting `rowsWritten` and `rowsWrittenLimit` into an embedded struct within `tableWriterBase`?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145934)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @yuzefovich)

___
*[`-- commits` line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNvXsPBm9IjIB-2gbB:-O_e2gYs0wDRzAiOSQ8Q:bu8zngk):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: technically, the first commit fixes the issue, no?
</blockquote></details>

I think you're right. Fixed it.
___
*[`pkg/sql/cancel_queries.go` line 20 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPWm0f6B9sR4o6TZc8J:-O_e7_h-342jGiRo1KTO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f444ed5b4af82f9013dde21369ba6567091cc01/pkg/sql/cancel_queries.go#L20)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I see, thanks.
</blockquote></details>

Done.
___
*[`pkg/sql/delete.go` line 198 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNryPdDKkr_jzb-YCo:-O_edp7c69bjdcFsVRTS:b-rlh8kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/70d2bfe5d60f388e79fe1be6f86c9c73a7e70277/pkg/sql/delete.go#L198)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I'm a bit confused about this setup. `r.td.row` will include the necessary KV requests to perform the deletion into `r.td.b`, but that KV batch will not be sent yet to the KV layer, which happens only in `deleteSwapNode.BatchedNext` via `td.finalize` call, so we shouldn't be seeing the ConditionFailedError here if it occurs (the only errors we should be seeing here are the encoding ones).

I think if the final swap PR is picked up on top of this change, the tests there would fail.

Perhaps just remove this comment and reset the counter to zero where we previously modified `rowsWritten`?
</blockquote></details>

Oh, great catch. I added a `reset` method that is now called in the event of a ConditionFailedError.
___
*[`pkg/sql/delete_swap.go` line 106 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNt2P970SNR483dV6d:-O_edcUy0X4eNp6u8VtN:b-4xf7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/delete_swap.go#L106)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: just double-checking: `rowsWritten` will now only be used for ensuring that we don't exceed `rowsWrittenLimit`? Perhaps, it'd be good to highlight that fact somehow? Maybe a comment, or extracting `rowsWritten` and `rowsWrittenLimit` into an embedded struct within `tableWriterBase`?
</blockquote></details>

Yes, that's right. I added a comment to the `rowsWritten` field.
___
*[`pkg/sql/plan_node_to_row_source.go` line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNwhZLCV_jh3YKzXnV:-O_e64Yo7w1CRx2GziXx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bce8048de0d4911c864d83ac29dd621a7c0ecce/pkg/sql/plan_node_to_row_source.go#L74)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think it should be safe to use a global singleton for the single INT type to avoid an allocation here. Perhaps leave a TODO to look into this.
</blockquote></details>

Done.
___
*[`pkg/sql/tablewriter.go` line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNtzDsAFYMAcNEu43M:-O_e6ZBT3qcsYHc9BLAh:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/tablewriter.go#L224)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think we can now remove `lastBatchSize` field?
</blockquote></details>

Good point, done.
___
*[`pkg/sql/colexec/columnarizer.go` line 220 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNxNRp3tGpqpyYfZj6:-O_e6yK2AObRxAnsO-U7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bce8048de0d4911c864d83ac29dd621a7c0ecce/pkg/sql/colexec/columnarizer.go#L220)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: something like `s/"fast-path" enabled planNodeToRowSource wrapper/planNodeToRowSource wrapper in "rows affected" mode/`.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/explain_analyze` line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OWNn1IKACuBXnuGNWLi:-O_e7AJJDXO8z-ZLk_Kh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bce8048de0d4911c864d83ac29dd621a7c0ecce/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze#L207)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this comment is now stale, maybe reword it as "Regression test for now showing execution stats for mutation nodes with RETURNING clause."?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Great work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @yuzefovich reviewed 6 of 29 files at r5, 23 of 23 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-:-O_gE_7d7AgMiO6tk9NC:b-4g32ek)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice clean up!

> @mgartner reviewed 3 of 24 files at r1, 2 of 29 files at r3, 29 of 29 files at r5, 23 of 23 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-:-O_luq352XIX8Kj-t-Ds:bjlvuu3)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/insert.go` line 274 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-OPqmyFe52rWBKUXhZxn:-O_lqFZJ7ml1mXPqLD1M:b95m4h9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3621cc06adfbcecf54c1b49e596bbda23934c599/pkg/sql/insert.go#L274)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

In the [old code path](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/opt_exec_factory.go#L1923-L1934), there were two cases:
1. We need rows from the mutation node, in which case we add a [spoolNode](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/spool.go#L16-L21) on top of it.
2. We don't need rows, in which case we add a [rowCountNode](https://github.com/cockroachdb/cockroach/blob/597bddc2323cf9441ec1182ee668e007454ef134/pkg/sql/plan_batch.go#L160-L172) on top of the mutation.

Both types of "wrapper" nodes ensured that the mutation is executed to completion upon the call to `startExec`, which happens before any rows are returned to the caller.
</blockquote></details>

Thanks for explaining!
___
*[`-- commits` line 20 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_lmKIe5kIGNc3iijrz:-O_lmKIe5kIGNc3iijs-:blf0cub):*
nit: "seme" and double periods at the end
___
*[`pkg/sql/delete_range.go` line 64 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_low0U4U2ljxD5QSU-:-O_low0U4U2ljxD5QSU0:b1pmk7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d76d97e89734f0e4f6de8bf7d7a69247934df8e4/pkg/sql/delete_range.go#L64)):*
> ```Go
> }
> 
> func (d *deleteRangeNode) rowsWritten() int64 {
> ```

nit: Should `rowsWritten` (maybe renamed, if you'd like) should be part of the mutation helper?
___
*[`pkg/sql/delete_swap.go` line 117 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_lpuELCRV96UAHT3Ss:-O_lpuELCRV96UAHT3St:bwxlfjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d76d97e89734f0e4f6de8bf7d7a69247934df8e4/pkg/sql/delete_swap.go#L117)):*
> ```Go
> 
> 	// Possibly initiate a run of CREATE STATISTICS.
> 	params.ExecCfg().StatsRefresher.NotifyMutation(d.run.td.tableDesc(), int(d.run.modifiedRowCount()))
> ```

nit: I don't think we need both `rowsWritten` methods and `modifiedRowCount` methods. They are the same thing, right?
___
*[`pkg/sql/insert.go` line 378 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_lu7Zt3pQ09V3en5ba:-O_lu7Zt3pQ09V3en5bb:b-r54s7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5a793da2430b8c58d7cf77c943fedc6fb5f17c1/pkg/sql/insert.go#L378)):*
> ```Go
> }
> 
> func (n *insertNode) returnsRowsAffected() bool {
> ```

nit: Should this be part of the mutation helper?
___
*[`pkg/sql/opaque.go` line 241 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_luHAn50tzX_V3gtu3:-O_luHAn50tzX_V3gtu4:b3icfzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5a793da2430b8c58d7cf77c943fedc6fb5f17c1/pkg/sql/opaque.go#L241)):*
> ```Go
> 		return p.Inspect(ctx, n)
> 	case *tree.MoveCursor:
> 		return p.MoveCursor(ctx, &n.CursorStmt)
> ```

Was this a bug?
___
*[`pkg/sql/plan_node_output_helper.go` line 20 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_lrf818RxyMoR-zRzt:-O_lrf818RxyMoR-zRzu:b-h30ajt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d76d97e89734f0e4f6de8bf7d7a69247934df8e4/pkg/sql/plan_node_output_helper.go#L20)):*
> ```Go
> 	// rowsNeeded is set to true if the mutation operator needs to return the rows
> 	// that were affected by the mutation.
> 	rowsNeeded bool
> ```

It's kinda weird to have `rowsNeeded` here when it's only used by structs that embed `mutationOutputHelper`. During execution, I believe this is equivalent to `rows != nil`, but `rowsNeeded` is set by the plan node constructors to communicate that rows are needed. So this is related to the problem that we've talked about where the distinct between planning data and working data is not clear in these planNodes.
___
*[`pkg/sql/plan_node_output_helper.go` line 47 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_ls0IHEf0YwMS2QW-o:-O_ls0II0nMrYbi2_6JL:b-q9rjus) ([raw file](https://github.com/cockroachdb/cockroach/blob/d76d97e89734f0e4f6de8bf7d7a69247934df8e4/pkg/sql/plan_node_output_helper.go#L47)):*
> ```Go
> 	*h = mutationOutputHelper{
> 		rowsNeeded: h.rowsNeeded,
> 		rows:       h.rows,
> ```

nit: I think there's more benefits to struct embedding when the API is defined through methods and we're not reaching into fields of substructs. At some point it would be nice to have an `Init` method to set `rows` and other fields as needed. This is already a great improvement though!


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: remove unnecessary nodes and interfaces related to mutation planNodes

This commit adds two new helper structs for use in mutation nodes and
other planNodes that need to output the number of affected rows:
`mutationOutputHelper` and `rowsAffectedOutputHelper`. The former is
now embedded into the mutation planNodes apart from `deleteRangeNode`,
which embeds `rowsAffectedOutputHelper` instead.

This commit also removes `spoolNode`, `serializeNode`, and `rowCount`
node, with `mutationOutputHelper` replacing the work they used to do.
This also allowed the removal of the `batchedPlanNode` interface and
its methods.

These changes simplify some factory logic, and should simplify the
process of moving from `planNode` implementations to processors in
the future.

This change also fixes #139378, where the `spoolNode` added for a mutation
with a RETURNING clause would break some of the EXPLAIN ANALYZE stats.

Fixes #139378

Release note (bug fix): Previously, CockroachDB would omit execution
statistics in EXPLAIN ANALYZE output for mutation nodes when a RETURNING
clause was used. The bug has been present since before v21.1 and is now fixed.

DrewKimball committed:
sql: change SetZoneConfig to return type Ack

This commit changes the return type for `tree.SetZoneConfig` to `Ack`
from `RowsAffected`. This mirrors similar statements, and the row count
has been broken (always zero) ever since this statement was implemented
in the declarative schema changer, anyway. Removing the need to track
row count allows a few related functions to be simplified.

Epic: None

Release note: None

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145934)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`-- commits` line 20 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_lmKIe5kIGNc3iijrz:-OaMH1PE9shFe_mKyopX:b-896fix):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "seme" and double periods at the end
</blockquote></details>

Done.
___
*[`pkg/sql/delete_range.go` line 64 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_low0U4U2ljxD5QSU-:-OaMIFVC1O6yVYE-KLla:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d76d97e89734f0e4f6de8bf7d7a69247934df8e4/pkg/sql/delete_range.go#L64)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Should `rowsWritten` (maybe renamed, if you'd like) should be part of the mutation helper?
</blockquote></details>

Done.
___
*[`pkg/sql/delete_swap.go` line 117 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_lpuELCRV96UAHT3Ss:-OaMIZbs4qX68YWzrtyV:b-p55j4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d76d97e89734f0e4f6de8bf7d7a69247934df8e4/pkg/sql/delete_swap.go#L117)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I don't think we need both `rowsWritten` methods and `modifiedRowCount` methods. They are the same thing, right?
</blockquote></details>

`rowsWritten` exposes the row count on the `mutationPlanNode` interface and is used for metrics, while `modifiedRowCount` is used internally to give mutation operators access to their internally tracked row count (including to implement `rowsWritten`. We could probably flatten out this nesting and make the helper directly implement the needed interface, but it doesn't seem worth it given these things are going to be significantly refactored in the future, anyway.
___
*[`pkg/sql/opaque.go` line 241 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_luHAn50tzX_V3gtu3:-OaMJpqUARBzWiriZ05C:bw7afbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5a793da2430b8c58d7cf77c943fedc6fb5f17c1/pkg/sql/opaque.go#L241)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Was this a bug?
</blockquote></details>

I believe we were previously relying on the row count fast-path logic to check the type of AST node, and convert from cursor output rows to row count for a `MOVE` statement. With this change, the `MOVE` statement now directly returns a row count, so it has to have its own implementation.
___
*[`pkg/sql/plan_node_output_helper.go` line 20 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_lrf818RxyMoR-zRzt:-OaMEjTMBsBLfxOEP40c:b-ehfxgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d76d97e89734f0e4f6de8bf7d7a69247934df8e4/pkg/sql/plan_node_output_helper.go#L20)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It's kinda weird to have `rowsNeeded` here when it's only used by structs that embed `mutationOutputHelper`. During execution, I believe this is equivalent to `rows != nil`, but `rowsNeeded` is set by the plan node constructors to communicate that rows are needed. So this is related to the problem that we've talked about where the distinct between planning data and working data is not clear in these planNodes.
</blockquote></details>

That's a good point. I've moved `rowsNeeded` back to the `.*Run` structs. There'll still be quite a bit of work separating out planning-only fields, but at least this way this PR isn't making things harder.
___
*[`pkg/sql/plan_node_output_helper.go` line 47 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/145934#-O_ls0IHEf0YwMS2QW-o:-OaM0lP3AsNzYN4z7a2Q:b-brsd3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d76d97e89734f0e4f6de8bf7d7a69247934df8e4/pkg/sql/plan_node_output_helper.go#L47)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: I think there's more benefits to struct embedding when the API is defined through methods and we're not reaching into fields of substructs. At some point it would be nice to have an `Init` method to set `rows` and other fields as needed. This is already a great improvement though!
</blockquote></details>

I'll leave a TODO and revisit once the next phase of refactoring (converting to the Processor model) is complete.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFYRs!

bors r+
